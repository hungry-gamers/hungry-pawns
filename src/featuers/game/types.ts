export type PawnSize = 'small' | 'medium' | 'big'

export interface Pawn {
  size: PawnSize
  playerId: string
}

export type Move = {
  size: PawnSize
  mode: 'pawn' | 'shield'
}

export type Location = { rowIndex: number; columnIndex: number }

export type ApplyShieldPayload = Location

export type SpecialMoveName = 'shield'

export type Sequence = `${string}/${number}-${number}/${number}-${number}/${number}-${number}`

export type Cell = { pawn: Pawn | null; shield: { activeInTurn: number; appliedBy: string } }
export type Board = Cell[][]
export type Player = {
  pawns: Record<PawnSize, number>
  capturedPawnsCounter: number
  arePawnsLocked: boolean
  specialMoves: SpecialMoveName[]
}

export type PlayerPayload = { id: string; pawns: Player['pawns'] }

export type PutPawnPayload = { pawnSize: PawnSize } & Location

export type Game = {
  status: 'not-initialized' | 'pregame' | 'in-progress' | 'finished'
  board: Board
  currentPlayerId: string
  turns: Record<number, { playerId: string; move: PutPawnPayload | ApplyShieldPayload }>
  winner: string | undefined
  allowedPawns: PawnSize[]
  sequences: Sequence[]
}
