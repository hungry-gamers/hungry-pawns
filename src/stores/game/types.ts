export type PawnSize = 'small' | 'medium' | 'big'

export interface Pawn {
  size: PawnSize
  playerId: string
}

export type Cell = Pawn | null
export type Board = (Cell | null)[][]
export type Player = {
  id: string
  pawns: {
    small: number
    medium: number
    big: number
  }
}

export type PutPawnPayload = { pawnSize: PawnSize; rowIndex: number; columnIndex: number }

export type Game = {
  status: 'not-initialized' | 'pregame' | 'in-progress' | 'finished'
  board: Board
  currentPlayerId: string
  pawns: Record<string, Record<PawnSize, number>>
  pawnsLockedBy: string[]
  turns: Record<number, { playerId: string; move: PutPawnPayload }>
  potentialWinner: string | undefined
  eatenPawnsCounter: Record<string, number>
}

type WinCondition = 'captured-line' | 'pawns-eaten' | 'balance' | 'dominance'

export type WinData = { winnerId: string; winCondition: WinCondition } | { winnerId: undefined }
