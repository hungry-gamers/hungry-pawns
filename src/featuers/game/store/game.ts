import { reactive } from 'vue'
import { defineStore } from 'pinia'
import * as GameT from '@/featuers/game/types.ts'
import * as PlayersT from '@/featuers/players/types'
import * as GameService from '@/featuers/game/services/game.service.ts'
import { usePlayersStore } from '@/featuers/players/store/players.ts'
import { SpecialEffectsSequences } from '@/featuers/game/services/game.service.ts'

export const useGameStore = defineStore('game', () => {
  const playersStore = usePlayersStore()

  const state = reactive<GameT.Game>({
    status: 'not-initialized',
    board: GameService.createBoard(),
    currentPlayerId: '',
    turns: {},
    winner: undefined,
    allowedPawns: ['small'],
    sequences: [],
  })

  const initiateGame = (players: PlayersT.PlayerPayload[]) => {
    state.status = 'pregame'
    state.board = GameService.createBoard()
    state.currentPlayerId = players[0].id
    state.allowedPawns = ['small']
    state.sequences = []
    state.winner = undefined

    playersStore.createPlayers(players)
  }

  const lockPawns = (playerId: string, pawns: Record<GameT.PawnSize, number>) => {
    if (playersStore.lockPawns(playerId, pawns)) {
      state.status = 'in-progress'
    }
  }

  const getLastTurn = () => {
    return Math.max(...Object.keys(state.turns).map(Number), 0)
  }

  const getCurrentTurn = () => {
    return getLastTurn() + 1
  }

  const manageAllowedPawns = () => {
    const lastTurn = getLastTurn()
    const rejectConditions = [
      lastTurn < GameService.UNLOCK_MEDIUM_PAWNS_TURN,
      state.allowedPawns.length === 2 && lastTurn < GameService.UNLOCK_BIG_PAWNS_TURN,
      lastTurn > GameService.UNLOCK_BIG_PAWNS_TURN,
    ]

    if (rejectConditions.some(Boolean)) return
    if (lastTurn === GameService.UNLOCK_MEDIUM_PAWNS_TURN) state.allowedPawns.push('medium')
    if (lastTurn === GameService.UNLOCK_BIG_PAWNS_TURN) state.allowedPawns.push('big')
  }

  const nextTurn = (payload?: GameT.PutPawnPayload | GameT.ApplyShieldPayload) => {
    playersStore.decreaseParalyzeProtection(state.currentPlayerId)
    state.turns[getLastTurn() + 1] = { playerId: state.currentPlayerId, move: payload || 'skip' }
    state.currentPlayerId = playersStore
      .getPlayers()
      .find((player) => player !== state.currentPlayerId)!

    manageAllowedPawns()
  }

  const checkInstantWin = () => {
    const CAPTURED_PAWNS_FOR_INSTANT_WIN = 5
    const player = playersStore.getPlayer(state.currentPlayerId)

    return player.capturedPawnsCounter === CAPTURED_PAWNS_FOR_INSTANT_WIN
  }

  const isLineCaptureWin = (): string | undefined => {
    const sequences = GameService.findWinningLine(state.board)

    if (state.sequences.length > 0) {
      const sequence = state.sequences.find((cell) => sequences.includes(cell))

      if (sequence) {
        const [playerId] = sequence.split('/')
        return playerId
      }
    }

    state.sequences = sequences

    return undefined
  }

  const skipTurn = () => {
    const result = playersStore.updateSkippedTurnsCount(state.currentPlayerId)
    const opponentId = playersStore.getOpponentId(state.currentPlayerId)

    const penalties: Record<number, () => void> = {
      1: () => playersStore.manipulatePawnAmount(opponentId, 'small', 1),
      2: () => playersStore.manipulatePawnAmount(opponentId, 'medium', 1),
      3: () => playersStore.manipulatePawnAmount(opponentId, 'big', 1),
      4: () => playersStore.addSpecialMove(opponentId, 'shield'),
      5: () => playersStore.addSpecialMove(opponentId, 'drop'),
    }

    penalties[result]()

    playersStore.removeParalyzeEffect(state.currentPlayerId)

    nextTurn()
  }

  const checkWinner = (payload: GameT.MovePayload) => {
    const isInstantWin = checkInstantWin()
    state.winner = isInstantWin ? state.currentPlayerId : isLineCaptureWin()

    if (!isInstantWin) nextTurn(payload)
    if (state.currentPlayerId === state.winner) {
      state.status = 'finished'
    }
  }

  const isCellProtected = ({ rowIndex, columnIndex }: GameT.Location) => {
    return state.board[rowIndex][columnIndex].shield.activeInTurn === getCurrentTurn()
  }

  const isPutPawnAllowed = (payload: GameT.PutPawnPayload) => {
    const player = playersStore.getPlayer(state.currentPlayerId)
    const { rowIndex, columnIndex, pawnSize } = payload
    const isPawnAvailable = player.pawns[payload.pawnSize] > 0
    const isPawnSizeAllowed = state.allowedPawns.includes(pawnSize)

    return (
      isPawnAvailable &&
      isPawnSizeAllowed &&
      !isCellProtected({ rowIndex, columnIndex }) &&
      !playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive
    )
  }

  const isDropPawnAllowed = (payload: GameT.DropOpponentPawnPayload) => {
    const { rowIndex, columnIndex } = payload
    const isPlayerPawn = state.board[rowIndex][columnIndex].pawn?.playerId === state.currentPlayerId

    return (
      !isCellProtected({ rowIndex, columnIndex }) &&
      !isPlayerPawn &&
      !playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive
    )
  }

  const dropOpponentPawn = (payload: GameT.MovePayload) => {
    if (!isDropPawnAllowed(payload)) return
    if (playersStore.useSpecialMove(state.currentPlayerId, 'drop') === 'not-allowed') return

    const { rowIndex, columnIndex } = payload
    state.board[rowIndex][columnIndex].pawn = null

    checkWinner(payload)
  }

  const paralyzePlayer = () => {
    const opponentId = playersStore.getOpponentId(state.currentPlayerId)
    const result = SpecialEffectsSequences.paralyze.every(([row, column]) => {
      return state.board[row][column].pawn?.playerId === state.currentPlayerId
    })

    if (!result || !playersStore.canPlayerBeParalyzed(opponentId)) return

    playersStore.paralyzePlayer(playersStore.getOpponentId(state.currentPlayerId))
  }

  const putPawn = (payload: GameT.PutPawnPayload) => {
    const { rowIndex, columnIndex, pawnSize } = payload
    const cellPawn = state.board[rowIndex][columnIndex].pawn

    if (!isPutPawnAllowed(payload) || state.status !== 'in-progress') return

    if (cellPawn) {
      if (!GameService.isPawnBigger(pawnSize, cellPawn.size)) return
      if (cellPawn.playerId !== state.currentPlayerId) {
        playersStore.updatePlayerCapturedPawns(state.currentPlayerId)
      }
    }

    state.board[rowIndex][columnIndex].pawn = {
      size: pawnSize,
      playerId: state.currentPlayerId,
    }
    playersStore.manipulatePawnAmount(state.currentPlayerId, pawnSize, -1)

    paralyzePlayer()
    checkWinner(payload)
  }

  const applyShield = (payload: GameT.ApplyShieldPayload) => {
    if (playersStore.getPlayer(state.currentPlayerId).paralyzed.isActive) return
    if (playersStore.useSpecialMove(state.currentPlayerId, 'shield') === 'not-allowed') return
    const { rowIndex, columnIndex } = payload

    state.board[rowIndex][columnIndex].shield.appliedBy = state.currentPlayerId
    state.board[rowIndex][columnIndex].shield.activeInTurn = getLastTurn() + 2

    checkWinner(payload)
  }

  return {
    state,
    initiateGame,
    putPawn,
    lockPawns,
    applyShield,
    getCurrentTurn,
    isLineCaptureWin,
    dropOpponentPawn,
    skipTurn,
  }
})
