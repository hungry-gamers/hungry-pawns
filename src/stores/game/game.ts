import { reactive } from 'vue'
import { defineStore } from 'pinia'
import type { Game, PutPawnPayload, Player, PawnSize, Cell } from '@/stores/game/types.ts'
import { AMOUNT_OF_EATEN_PAWNS_FOR_INSTANT_WIN } from '@/utils/constants.ts'

export const useGameStore = defineStore('game', () => {
  const state = reactive<Game>({
    status: 'not-initialized',
    board: Array.from({ length: 3 }, () => Array(3).fill(null)),
    currentPlayerId: '',
    pawns: {},
    pawnsLockedBy: [],
    turns: {},
    eatenPawnsCounter: {},
    potentialWinner: undefined,
    allowedPawns: ['small'],
  })

  const initiateGame = (players: Player[]) => {
    state.status = 'pregame'
    state.board = Array.from({ length: 3 }, () => Array(3).fill(null))
    state.currentPlayerId = players[0].id
    state.allowedPawns = ['small']

    players.forEach((player) => {
      state.pawns[player.id] = { ...player.pawns }
      state.eatenPawnsCounter[player.id] = 0
    })
  }

  const getPlayers = (): string[] => {
    return Object.keys(state.pawns ?? {})
  }

  const lockPawns = (playerId: string, pawns: Record<PawnSize, number>) => {
    if (state.pawnsLockedBy.includes(playerId)) return
    state.pawnsLockedBy.push(playerId)
    state.pawns[playerId].small = pawns.small
    state.pawns[playerId].medium = pawns.medium
    state.pawns[playerId].big = pawns.big
    const [player1Id, player2Id] = getPlayers()

    if (state.pawnsLockedBy.includes(player1Id) && state.pawnsLockedBy.includes(player2Id)) {
      state.status = 'in-progress'
    }
  }

  const isBigger = (size: PawnSize, pawnOnBoardSize: PawnSize) => {
    const order: PawnSize[] = ['small', 'medium', 'big']
    return order.indexOf(size) > order.indexOf(pawnOnBoardSize)
  }

  const isInstantWin = () => {
    return state.eatenPawnsCounter[state.currentPlayerId] === AMOUNT_OF_EATEN_PAWNS_FOR_INSTANT_WIN
  }

  const isLineCaptureWin = (): string | undefined => {
    const board = state.board

    const isWinningLine = (line: Cell[]) => {
      if (line.some((cell) => !cell)) return false
      const playerId = line[0]?.playerId
      return line.every((cell) => cell?.playerId === playerId)
    }

    for (let i = 0; i < 3; i++) {
      if (isWinningLine([board[i][0], board[i][1], board[i][2]])) return board[i][0]?.playerId
      if (isWinningLine([board[0][i], board[1][i], board[2][i]])) return board[0][i]?.playerId
    }

    if (isWinningLine([board[0][0], board[1][1], board[2][2]])) return board[0][0]?.playerId
    if (isWinningLine([board[0][2], board[1][1], board[2][0]])) return board[0][2]?.playerId

    return undefined
  }

  const putPawn = (payload: PutPawnPayload) => {
    const { rowIndex, columnIndex, pawnSize } = payload
    const currentPlayer = state.currentPlayerId
    const cellPawn = state.board[rowIndex][columnIndex]
    const lastTurn = Math.max(...Object.keys(state.turns).map(Number), 0)
    const moveNotAllowed =
      !state.pawns[currentPlayer][pawnSize] || !state.allowedPawns.includes(payload.pawnSize)

    if (moveNotAllowed || state.status !== 'in-progress') return

    if (cellPawn) {
      if (!isBigger(pawnSize, cellPawn.size)) return
      if (cellPawn.playerId !== state.currentPlayerId) {
        state.eatenPawnsCounter[currentPlayer]++
      }
    }

    state.board[rowIndex][columnIndex] = { size: pawnSize, playerId: currentPlayer }
    state.pawns[currentPlayer][pawnSize]--
    state.turns[lastTurn + 1] = { playerId: currentPlayer, move: payload }
    const turnsPlayed = Object.keys(state.turns).length

    if (state.allowedPawns.length === 1 && turnsPlayed === 4) {
      state.allowedPawns.push('medium')
    }
    if (state.allowedPawns.length === 2 && turnsPlayed === 8) {
      state.allowedPawns.push('big')
    }

    state.potentialWinner = isInstantWin() ? state.currentPlayerId : isLineCaptureWin()

    if (!isInstantWin()) {
      state.currentPlayerId = Object.keys(state.pawns).find((player) => player !== currentPlayer)!
    }

    if (state.currentPlayerId === state.potentialWinner) {
      state.status = 'finished'
    }
  }

  return {
    state,
    initiateGame,
    putPawn,
    getPlayers,
    lockPawns,
    isLineCaptureWin,
  }
})
