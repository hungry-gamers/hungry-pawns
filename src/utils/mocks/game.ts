import type { Player } from '@/stores/game/types.ts'

export const players: Player[] = [
  { id: '1', pawns: { small: 3, medium: 3, big: 3 } },
  { id: '2', pawns: { small: 3, medium: 3, big: 3 } },
]
