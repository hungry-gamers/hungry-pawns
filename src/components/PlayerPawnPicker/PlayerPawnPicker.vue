<script setup lang="ts">
import PawnAmountCounter from '@/components/PawnAmountCounter/PawnAmountCounter.vue'
import { ref, computed } from 'vue'
import { MAXIMUM_PAWNS_PER_PLAYER } from '@/utils/constants.ts'
import type { PawnSize } from '@/featuers/game/types.ts'
import { usePlayersStore } from '@/featuers/players/store/players.ts'
import { useGameStore } from '@/featuers/game/store/game.ts'

const props = defineProps<{ playerId: string }>()

const { lockPawns } = useGameStore()
const { state } = usePlayersStore()

const pawns = ref({ ...state.players[props.playerId]?.pawns })

const isLocked = computed(() => state.players[props.playerId]?.arePawnsLocked)
const totalPawns = computed(() =>
  Object.values(pawns.value as Record<string, number>).reduce((sum, current) => sum + current, 0),
)

const isIncreaseButtonDisabled = computed(() => totalPawns.value === MAXIMUM_PAWNS_PER_PLAYER)
const canGetMorePawns = computed(() => totalPawns.value < MAXIMUM_PAWNS_PER_PLAYER)

const onAmountChange = (payload: { amount: number; size: PawnSize }) => {
  pawns.value[payload.size] = payload.amount
}
</script>

<template>
  <div>
    <span>Player: {{ playerId }}</span>
    <div v-if="isLocked">
      Your pawns: small: {{ pawns.small }} medium: {{ pawns.medium }} big: {{ pawns.big }}
    </div>
    <div v-else>
      <PawnAmountCounter
        size="small"
        :amount="pawns.small"
        :is-increase-disabled="isIncreaseButtonDisabled"
        @on-amount-changed="onAmountChange"
      />
      <PawnAmountCounter
        size="medium"
        :amount="pawns.medium"
        :is-increase-disabled="isIncreaseButtonDisabled"
        @on-amount-changed="onAmountChange"
      />
      <PawnAmountCounter
        size="big"
        :amount="pawns.big"
        :is-increase-disabled="isIncreaseButtonDisabled"
        @on-amount-changed="onAmountChange"
      />
    </div>

    <button
      @click="lockPawns(playerId, pawns)"
      :disabled="isLocked || canGetMorePawns"
      data-test-id="lock-pawns-button"
    >
      Lock
    </button>
  </div>
</template>

<style scoped></style>
