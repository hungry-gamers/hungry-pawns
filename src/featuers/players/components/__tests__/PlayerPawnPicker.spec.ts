import { describe, expect, it, vi } from 'vitest'
import PlayerPawnPicker from '../PlayerPawnPicker.vue'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { useGameStore } from '@/featuers/game/store/game.ts'
import { players } from '@/utils/mocks/game.ts'

describe('PlayerPawnPicker', () => {
  const pinia = createTestingPinia({
    createSpy: vi.fn,
    stubActions: false,
  })

  const store = useGameStore()
  const spy = vi.spyOn(store, 'lockPawns')
  store.initiateGame(players)

  const wrapper = mount(PlayerPawnPicker, {
    props: {
      playerId: '1',
    },
    global: {
      plugins: [pinia],
    },
  })

  it('should render properly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should lock pawns', async () => {
    const button = wrapper.find('[data-test-id="lock-pawns-button"]')
    await button.trigger('click')

    expect(spy).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalledWith('1', { small: 3, medium: 3, big: 3 })
  })
})
