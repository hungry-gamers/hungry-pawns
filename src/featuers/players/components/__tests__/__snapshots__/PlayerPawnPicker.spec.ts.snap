// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerPawnPicker > should render properly 1`] = `
"<div><span>Player: 1</span>
  <div>
    <div><span>small</span><button data-test-id="decrease-button"> - </button><span>3</span><button disabled="" data-test-id="increase-button"> + </button></div>
    <div><span>medium</span><button data-test-id="decrease-button"> - </button><span>3</span><button disabled="" data-test-id="increase-button"> + </button></div>
    <div><span>big</span><button data-test-id="decrease-button"> - </button><span>3</span><button disabled="" data-test-id="increase-button"> + </button></div>
  </div><button data-test-id="lock-pawns-button"> Lock </button>
</div>"
`;
